lib/os/fdtable.c

```diff
@@ -65,6 +65,11 @@ static int z_fd_unref(int fd)
 {
 	int old_rc = atomic_dec(&fdtable[fd].refcount);
 
+	/* Reference counter must be checked to avoid decrement refcount below
+	 * zero causing file descriptor leak.
+	 */
+	atomic_cas(&fdtable[fd].refcount, -1, 0);
```

Comment:
Much better now, but there is still a possibility that the refcount value is changed between a call to `atomic_dec()` and `atomic_cas()`. Not sure how to solve that using just the atomic API.

Reply:
Yes. There is such possibility. However, it should not be harmful, because `fdtable[fd].refcount == -1` will be only for a short while. There is a small chance that between `atomic_dec` and `atomic_cas` there will be concurrent call of `z_reserve_fd` and file descriptor will be still allocated.

To solve this problem we probably need new atomic method. Something like: `atomic_dec_if_not_zero`. However, solution above is IMO good enough and thread safe.

Reply:
I just remembered that we have this kind of thing in net_pkt for unreffing, so we could do it like this here too
```
atomic_val_t ref;

do {
	ref = atomic_get(&fdtable[fd].refcount);
	if (!ref) {
		return;
	}
} while (!atomic_cas(&fdtable[fd].refcount, ref, ref - 1));
```

Reply:
Yeah. Good point. Please take a look now.